src/llvm-cpufeatures.cpp

```diff
@@ -0,0 +1,110 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+//
+// Lower intrinsics that expose subtarget information to the language. This makes it
+// possible to write code that changes behavior based on, e.g., the availability of
+// specific CPU features.
+//
+// The following intrinsics are supported:
+// - julia.cpu.have_fma: returns 1 if the platform supports hardware-accelerated FMA
+//
+// XXX: can / do we want to make this a codegen pass to enable querying TargetPassConfig?
+
+#include "llvm-version.h"
+
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/PassManager.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/Target/TargetMachine.h>
+#include <llvm/Support/Debug.h>
+
+#include "julia.h"
+
+#define DEBUG_TYPE "cpufeatures"
+
+using namespace llvm;
+
+extern TargetMachine *jl_TargetMachine;
+
+namespace {
+
+static void lowerHaveFMA(Function &F, Instruction *I) {
+    Triple TheTriple = Triple(jl_TargetMachine->getTargetTriple());
+
+    Attribute CPUAttr = F.getFnAttribute("target-cpu");
+    Attribute FSAttr = F.getFnAttribute("target-features");
+
+    StringRef CPU =
+        CPUAttr.isValid() ? CPUAttr.getValueAsString() : jl_TargetMachine->getTargetCPU();
+    StringRef FS =
+        FSAttr.isValid() ? FSAttr.getValueAsString() : jl_TargetMachine->getTargetFeatureString();
+
+    if (TheTriple.getArch() == Triple::x86_64 && FS.find("+fma") != StringRef::npos)
```

Comment:
are there any cpus that support fma4 but not fma3?

Reply:
Bulldozer.

Reply:
Didn't realized that you just searched for the fma prefix, while this might work right now it'll also catch cases where that prefix appears in unrelated features.

Also it seems that on arm there's a way to enable vfp4 for single precision only and not for double precision so I think even if you don't want to implement it for arm right now, the intrinsic should be updated to support that (I assume sth like adding `.f32` and `f64` suffix)

Reply:
Does LLVM have helpers to inspect the feature string instead of doing string matching?

Reply:
I presume you can't use [`jl_test_cpu_feature`](https://github.com/JuliaLang/julia/blob/0a4338b3178341762faac5cec63b295eb1fdd24f/src/processor.h#L126) because that'd test the feature on the host and not the target?

Reply:
> Does LLVM have helpers to inspect the feature string instead of doing string matching?

Cut the string and find?

> I presume you can't use jl_test_cpu_feature because that'd test the feature on the host and not the target?

Correct. `jl_test_cpu_feature` should only be used by runtime functions and not the compiler.